### PR TITLE
Multi API support for NPU and DATAPLANE.

### PR DESCRIPTION
**Pull request by anton7811**
1. Infrastructure changes:
- All test infrastructure now can be defined in a single JSON file.
- Added a possibility to define multiple DUTs.

2. DUT-related changes:
- Added Thrift client support
- Added a possibility to switch between Redis and Thrift protocols for SAI RPC in JSON testbed configuration.
- Added new DUT type for DASH - DPU.

3. Dataplane changes:
- Decoupled PTF init from conftest.py and moved to a separate module.
- Added snappi module with initial traffic control methods.

Known issues:
- Redis tests stability (replacing oid in some tests is not valid)
- Dynamic Thrift call attribute typing (currently hard coded)

### CODE REVIEW

File: Dockerfile.server [hunk 1]
```diff
@@ -114,17 +114,8 @@ COPY scripts/redis-cmd-listener.py   /sai-challenger/scripts/redis-cmd-listener.
...
-COPY cli/main.py         /sai-challenger/cli/main.py
-COPY common/setup.py     /sai-challenger/common/setup.py
-COPY common/__init__.py  /sai-challenger/common/__init__.py
-COPY common/sai.py       /sai-challenger/common/sai.py
-COPY common/sai_npu.py   /sai-challenger/common/sai_npu.py
-COPY common/sai_dataplane.py         /sai-challenger/common/sai_dataplane.py
-COPY topologies          /sai-challenger/topologies
-COPY .build/npu          /sai-challenger/npu
-COPY scripts/sai-cli-completion.sh   /sai-challenger/scripts/sai-cli-completion.sh
+ADD . /sai-challenger/
```
Comment (andriy-kokhan) [inline Dockerfile.server:117]: Why do we do this? Looks like all TCs will be installed into the server Docker which is not right...

File: common/sai.py [hunk 2]
```diff
@@ -1,180 +1,231 @@
...
+import logging
 import os
 import pytest
 
+from saichallenger.common.sai_abstractions import AbstractEntity
+from saichallenger.common.sai_client.sai_client import SaiClient
+from saichallenger.common.sai_data import SaiObjType
+
+
+class Sai(AbstractEntity):
```
Comment (andriy-kokhan) [inline common/sai.py:11]: The `AbstractEntity` contains `init_connections()` which is not related to SAI interface. So, `Sai` should not be inherited from `AbstractEntity`.

File: common/sai.py [hunk 3]
```diff
@@ -214,776 +265,80 @@ def get_obj_attr_type(sai_obj_type, sai_obj_attr):
...
-
-        if do_assert:
-            print(entry_status)
-            assert status[2] == 'SAI_STATUS_SUCCESS'
-            return status[2], entry_status
-
-        return status[2], entry_status
-
-    def get_by_type(self, obj, attr, attr_type, do_assert = True):
+    def get_by_type(self, obj, attr, attr_type):
```
Comment (andriy-kokhan) [inline common/sai.py:268]: Do we use it for Thrift client? I think its Redis specific and should be in sai_redis_client..

File: common/sai_client/sai_client.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+        raise NotImplementedError
+
+    def cleanup(self):
+        raise NotImplementedError
+
+    def set_loglevel(self, sai_api, loglevel):
+        raise NotImplementedError
+
+    # CRUD
+    def create(self, obj_type, *, key=None, attrs=None, do_assert=True):
```
Comment (andriy-kokhan) [inline common/sai_client/sai_client.py:18]: I think `*, ` here is excessive. Due to this, the TCs will be polluted with params names. Please remove here and below.

File: setups/sai_dpu_client_server.json [hunk 5]
```diff
@@ -0,0 +1,42 @@
+{"DPU": [
```
Comment (andriy-kokhan) [inline setups/sai_dpu_client_server.json:1]: Please confirm that a setup description file is optional and original approach to run TCs in standalone/client-server mode is supported.

File: setups/sai_dpu_client_server.json [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+  {
+    "alias": "dash",
+    "asic": "generic",
+    "target": "bmv2",
+    "type": null,
+    "sku": null,
```
Comment (andriy-kokhan) [inline setups/sai_dpu_client_server.json:7]: What's the right way to provide `sku` ?

File: common/sai_client/sai_thrift_client/sai_thrift_client.py [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+    def __del__(self):
+        self.thrift_transport.close()
+
+    @assert_status
+    def create(self, obj_type, *, key=None, attrs=()):
+        oid_or_status = self._operate('create', attrs=attrs, obj_type=obj_type, key=key)
+        return oid_or_status if key is None else key
+
+    @assert_status
+    def remove(self, *, oid=None, obj_type=None, key=None):
```
Comment (andriy-kokhan) [inline common/sai_client/sai_thrift_client/sai_thrift_client.py:51]: We should be able to get `obj_type` by `oid`. So, why do we need this extra param? Same for `key`..

File: Dockerfile [hunk 8]
```diff
@@ -116,6 +116,16 @@ RUN echo ". /sai-challenger/scripts/sai-cli-completion.sh" >> /root/.bashrc
 
 RUN pip3 install pytest pytest_dependency pytest-html
 
+
+# Deploy SAI Challenger CLI
+ADD . /sai-challenger/
+RUN ln -s /sai-challenger /usr/local/lib/python3.7/dist-packages/saichallenger
+RUN echo ". /sai-challenger/scripts/sai-cli-completion.sh" >> /root/.bashrc
```
Comment (andriy-kokhan) [inline Dockerfile:123]: Already done in L115

File: Dockerfile [hunk 9]
```diff
@@ -116,6 +116,16 @@ RUN echo ". /sai-challenger/scripts/sai-cli-completion.sh" >> /root/.bashrc
...
+
+# Install SAI Challenger CLI and the remote commands listener
+RUN pip3 install --editable /sai-challenger/common /sai-challenger
```
Comment (andriy-kokhan) [inline Dockerfile:126]: For a standalone mode this is done through supervisord.conf
https://github.com/opencomputeproject/SAI-Challenger/blob/beece67f68142de5b7486f5a839eacd7010fe2b2/configs/supervisord.conf#L25

File: common/sai_client/sai_redis_client/sai_redis_client.py [hunk 10]
```diff
@@ -0,0 +1,781 @@
...
+            attempts -= 1
+
+        if attempts == 0:
+            return []
+
+        # Remove spaces from the key string.
+        # Required by sai_serialize_route_entry() in sairedis.
+        obj = obj.replace(' ', '')
+
+        # Make redis-style OIDs
```
Comment (andriy-kokhan) [inline common/sai_client/sai_redis_client/sai_redis_client.py:745]: Why do we need this?

File: common/sai_client/sai_redis_client/sai_redis_client.py [hunk 11]
```diff
@@ -0,0 +1,781 @@
...
+        attrs = attrs.replace('oid:0x', '0x')
+        attrs = attrs.replace('0x', 'oid:0x')
+
+        self.r.lpush("ASIC_STATE_KEY_VALUE_OP_QUEUE", obj, attrs, op)
+        self.r.publish("ASIC_STATE_CHANNEL@1", "G")
+
+        status = []
+        attempts = self.attempts
+
+        #TODO: Handle 'vs' from the config
```
Comment (andriy-kokhan) [inline common/sai_client/sai_redis_client/sai_redis_client.py:757]: Is this from the original implementation? If "yes", why is this commented out?

File: common/sai_data.py [hunk 12]
```diff
@@ -0,0 +1,173 @@
+import json
+from enum import Enum
+
+
+# TODO: make it dynamicly generated from headers
+class SaiObjType(Enum):
```
Comment (andriy-kokhan) [inline common/sai_data.py:6]: The numbers should be aligned as it was originally in sai.py

File: common/sai_dataplane.py [hunk 13]
```diff
@@ -1,21 +1,69 @@
...
 from ptf import config
-import os
-import unittest
-import copy
 
+from sai_abstractions import AbstractEntity
+
+# TODO: Get rid of this
+# The default configuration dictionary for PTF_NN
+config_default = {
```
Comment (andriy-kokhan) [inline common/sai_dataplane.py:13]: I think this is PTF specific and should not be in common file.

File: dataplane/ptf_testutils.py [hunk 14]
```diff
@@ -0,0 +1,539 @@
...
+from scapy.layers.l2 import Ether
+
+import ptf.testutils as ptfTestutils
+
+default_step = 0.2
+default_timeout = 2
+default_n_timeout = 2
+
+
+class SnappiDataPlaneUtilsWrapper:
```
Comment (andriy-kokhan) [inline dataplane/ptf_testutils.py:15]: Why do we need it? How do we use it?

File: dataplane/snappi/sai_dataplane.py [hunk 15]
```diff
@@ -0,0 +1,334 @@
...
+            print("Warning: %s" % str(response.warnings))
+
+    @staticmethod
+    def get_capture_port_names(cfg):
+        """
+        Returns name of ports for which capture is enabled.
+        """
+        names = []
+        for cap in cfg.captures:
+            if cap._properties.get("port_names"):
```
Comment (ashutshkumr) [inline dataplane/snappi/sai_dataplane.py:82]: would suggest not using private members here; should use `if cap.port_names` instead.

File: dataplane/snappi/sai_dataplane.py [hunk 16]
```diff
@@ -0,0 +1,334 @@
...
+                        names.append(name)
+        return names
+
+    def set_config(self, return_time=False):
+        start_time = time.time()
+        logging.info('Setting config ...')
+        res = self.api.set_config(self.configuration)
+        assert self.api_results_ok(res), res
+        if len(res.warnings) > 0:
+            logging.info('Warnings in set_config : {}'.format(res.warnings))
```
Comment (ashutshkumr) [inline dataplane/snappi/sai_dataplane.py:94]: Couple of notes:
- Upon error, exception is raised. If this is not done across different implementation, then it's a bug in the implementation
- timers should be guarded by `try/finally` to guarantee execution
- Warnings could technically be logged inside a different func

Here's how we do it: https://github.com/open-traffic-generator/conformance/blob/dd628aa14c2ae5e4f4d8d4858b07a6acc50ce6d0/helpers/otg/otg.py#L113

File: dataplane/snappi/sai_dataplane.py [hunk 17]
```diff
@@ -0,0 +1,334 @@
...
+        if return_time:
+            return operation_time
+
+    def start_capture(self):
+        """ Start a capture which was already configured.
+        """
+        capture_names = self.get_capture_port_names(self.configuration)
+        logging.info(f"Starting capture on ports {str(capture_names)}")
+        cs = self.api.capture_state()
+        cs.state = cs.START
```
Comment (ashutshkumr) [inline dataplane/snappi/sai_dataplane.py:106]: would advice providing name of ports to start capture on

File: dataplane/snappi/sai_dataplane.py [hunk 18]
```diff
@@ -0,0 +1,334 @@
...
+        self.check_warnings(self.api.set_capture_state(cs))
+
+    def stop_capture(self):
+        capture_names = self.get_capture_port_names(self.configuration)
+        logging.info(f"Stopping capture on ports {str(capture_names)}")
+        cs = self.api.capture_state()
+        cs.state = cs.STOP
```
Comment (ashutshkumr) [inline dataplane/snappi/sai_dataplane.py:113]: would advice providing name of ports to start capture on

File: dataplane/snappi/sai_dataplane.py [hunk 19]
```diff
@@ -0,0 +1,334 @@
...
+
+        # Configure two ports where the location points to the port location:
+        # Ixia-C:       port.location = "localhost:5555"
+        # IxNetwork:    port.location = "<chassisip>;card;port"
+        # TRex:         port.location = "localhost:5555"
+        for pid, port in enumerate(self.exec_params['port_groups']):
+            location = port.get('location', f"localhost:{BASE_TENGINE_PORT+pid}")
+            self.configuration.ports.port(name=port[port['init']], location=location)
+
+        cap = self.configuration.captures.capture(name="c1")[-1]
```
Comment (ashutshkumr) [inline dataplane/snappi/sai_dataplane.py:42]: would suggest adding new ports/captures using newly added `add` construct.

https://github.com/open-traffic-generator/conformance/blob/main/features/flows/headers/udp/test_udp_header.py#L42

https://github.com/open-traffic-generator/conformance/blob/main/features/flows/headers/udp/test_udp_header.py#L49

File: dataplane/snappi/sai_dataplane.py [hunk 20]
```diff
@@ -0,0 +1,334 @@
...
+    def add_flow(self,
+                 name,
+                 packet_count=1,
+                 seconds_count=0,
+                 pps=10,
+                 force_pps=False
+                 ):
+
+        name = name or f"flow_{time.time()}"
+        flow = self.configuration.flows.flow(name=name)[-1]
```
Comment (ashutshkumr) [inline dataplane/snappi/sai_dataplane.py:214]: would suggest adding new flow using new `add` construct.

https://github.com/open-traffic-generator/conformance/blob/main/features/flows/headers/udp/test_udp_header.py#L52

File: dataplane/snappi/sai_dataplane.py [hunk 21]
```diff
@@ -0,0 +1,334 @@
...
+        return all([m.transmit == "stopped" for m in metrics])
+
+    def get_all_stats(self, print_output=True):
+        """
+        Returns all port and flow stats
+        """
+        print("Fetching all port stats ...")
+        request = self.api.metrics_request()
+        request.choice = request.PORT
+        request.port
```
Comment (ashutshkumr) [inline dataplane/snappi/sai_dataplane.py:178]: this can be replaced with `req.port.port_names = []`

similar suggestion for flow_metrics.

**Review by ashutshkumr — CHANGES_REQUESTED** [review]
Requested few changes in `dataplane/snappi/sai_dataplane.py`

File: dataplane/snappi/sai_dataplane.py (hunk 19 above)
Comment (anton7811) [reply dataplane/snappi/sai_dataplane.py:42]: So we will be able to capture per port? That's will be extremely helpful! Thanks!

Comment (ashutshkumr) [reply dataplane/snappi/sai_dataplane.py:42]: yes, indeed, you should be able to capture on per port basis.

File: dataplane/snappi/sai_dataplane.py (hunk 21 above)
Comment (anton7811) [reply dataplane/snappi/sai_dataplane.py]: fixed

File: dataplane/snappi/sai_dataplane.py [hunk 22]
```diff
@@ -0,0 +1,334 @@
...
+            return None
+
+        ether = flow.packet.add().ethernet
+        ether.dst.value = dst_mac
+        ether.src.value = src_mac
+        ether.ether_type.value = eth_type
+
+        # Setup increment
+        self.set_increment(ether.dst, dst_choice, dst_count, dst_mac, dst_step)
+        self.set_increment(ether.src, src_choice, src_count, src_mac, src_step)
```
[the remark this answers is not in the document]
Comment (anton7811) [reply dataplane/snappi/sai_dataplane.py]: fixed

File: dataplane/snappi/sai_dataplane.py (hunk 19 above)
Comment (anton7811) [reply dataplane/snappi/sai_dataplane.py:42]: Added to backlog.

File: dataplane/snappi/sai_dataplane.py (hunk 20 above)
Comment (anton7811) [reply dataplane/snappi/sai_dataplane.py:215]: Added to backlog.

File: dataplane/snappi/sai_dataplane.py (hunk 18 above)
Comment (anton7811) [reply dataplane/snappi/sai_dataplane.py:115]: fixed

File: dataplane/snappi/sai_dataplane.py (hunk 17 above)
Comment (anton7811) [reply dataplane/snappi/sai_dataplane.py:108]: Added to backlog

File: dataplane/snappi/sai_dataplane.py (hunk 16 above)
Comment (anton7811) [reply dataplane/snappi/sai_dataplane.py]: fixed

File: dataplane/snappi/sai_dataplane.py (hunk 15 above)
Comment (anton7811) [reply dataplane/snappi/sai_dataplane.py]: fixed

**Comment by andriy-kokhan** [comment]
The changes from this PR are included in https://github.com/opencomputeproject/SAI-Challenger/pull/26

**Pull request closed without merging**